db/commitlog/commitlog.cc

```diff
@@ -1392,83 +1393,84 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
 future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager::allocate_segment_ex(descriptor d, sstring filename, open_flags flags) {
     file_open_options opt;
     opt.extent_allocation_size_hint = max_size;
-    auto fut = do_io_check(commit_error_handler, [this, filename, flags, opt] () mutable {
-        auto fut = open_file_dma(filename, flags, opt);
-        if (cfg.extensions && !cfg.extensions->commitlog_file_extensions().empty()) {
-            for (auto * ext : cfg.extensions->commitlog_file_extensions()) {
-                fut = with_file_close_on_failure(std::move(fut), [ext, filename = std::move(filename), flags](file f) mutable {
-                   return ext->wrap_file(std::move(filename), f, flags).then([f](file nf) mutable {
-                       return nf ? nf : std::move(f);
-                   });
-                });
-            }
-        }
-        return fut;
-    });
 
-    return with_file_close_on_failure(std::move(fut), [this, d = std::move(d), filename = std::move(filename), flags] (file f) mutable {
-        f = make_checked_file(commit_error_handler, f);
-        // xfs doesn't like files extended betond eof, so enlarge the file
-        auto fut = make_ready_future<>();
-        // If file is opened with O_DSYNC, we should explicitly write zeros
-        // instead of just truncate/fallocate. Otherwise we get crappy
-        // behaviour.
+    file f;
+    std::exception_ptr ep;
+
+    try {
```

Comment:
BTW, if we don't l like try/catch, we could have a lambda with co-awaits and co-returns, which is then-wrapped(). I don't mind either style, but remember this discussion that try/catch is not welcome (I wonder how this policy fares now).

Reply:
Why would try-catch no be "welcome"? Then is a wrapped try-catch. Only thing to remember is that actual coroutine calls are not allowed in catch block (hence the deferred ep above), and of course neither are they in a destructor block (for both same reason and destructor signaure). 
But a try-catch is just a slimmer way to deal with all the endless wrapping done previously.

Reply:
I remember this discussion reflected in the style guide: http://docs.seastar.io/master/tutorial.html#exceptions-vs.-exceptional-futures

Reply:
That is for then(..)-style code. Once we're on coroutines, normal exception management applies, as once you do co_await and friends, all exceptional futures will be promoted to throws.